# Use `cellranger-atac` to align sequencing files
Refer to https://ucdavis-bioinformatics-training.github.io/2022-July-Advanced-Topics-in-Single-Cell-RNA-Seq-ATAC/data_reduction/ATAC to learn more about how to use cellranger-atac
## Docker image
Use `cumulusprod/cellranger-atac:2.1.0` to implement `cellranger-atac`
```
bsub -Is -q general-interactive -G compute-yeli -R 'rusage[mem=32GB]' -a 'docker(cumulusprod/cellranger-atac:2.1.0)' /bin/bash
```
After pulling down the docker image, add environment variables
```
export PATH=/software:/software/cellranger-atac-2.1.0:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH
```
To find environment variables for certain command/apps in certain docker images:
- Docker pull on your own device and execute the container.
- Run this command in the root environment: `echo $PATH`

## Reference
There are prebuilt human and mouse reference packages for use with Cell Ranger ATAC, which we will be using in this workshop. For other species, or to create a custom reference, one can use cellranger-atac mkref.
### Custom reference genome
The following code will generate a reference with `cellranger-atac mkref`. This takes a while. When using this code, ensure that your FASTA and GTF files are appropriate versions, downloading up to date files as necessary.
```
cd /path/to/project
cellranger-atac mkref \
   --config=path/to/config
```
### Download prebuild CellRanger reference genome
```
cd /path/to/data/
curl -O https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-arc-mm10-2020-A-2.0.0.tar.gz

# for hg38
# curl -O https://cf.10xgenomics.com/supp/cell-atac/refdata-cellranger-arc-GRCh38-2020-A-2.0.0.tar.gz

tar -xzvf refdata-cellranger-arc-mm10-2020-A-2.0.0.tar.gz
rm refdata-cellranger-arc-mm10-2020-A-2.0.0.tar.gz
```
## Organize custom sequencing data files
Rename the files generated by `fastq-dump` into R1, R2, R3, and I1.

```
GSE232482/
├── m1/
│   ├── sample1_L001_R1_001.fastq.gz
│   ├── sample1_L001_R2_001.fastq.gz
│   ├── sample1_L001_R3_001.fastq.gz
│   └── sample1_L001_I1_001.fastq.gz
│   
├── m2/
│   ├── sample2_L001_R1_001.fastq.gz
│   ├── sample2_L001_R2_001.fastq.gz
│   ├── sample2_L001_R3_001.fastq.gz
│   └── sample2_L001_I1_001.fastq.gz
│ 
├── f1/
│   ├── sample3_L001_R1_001.fastq.gz
│   ├── sample3_L001_R2_001.fastq.gz
│   ├── sample3_L001_R3_001.fastq.gz
│   └── sample3_L001_I1_001.fastq.gz
│   
└── f2/
    ├── sample4_L001_R1_001.fastq.gz
    ├── sample4_L001_R2_001.fastq.gz
    ├── sample4_L001_R3_001.fastq.gz
    └── sample4_L001_I1_001.fastq.gz
```
In SRA (Sequence Read Archive) data, R1, R2, R3, and I1 are commonly used to describe different types of read segments (reads) in high-throughput sequencing data files:
- R1: This is the first sequence read, usually a forward read from paired end sequencing. If it is single-end sequencing, there is only R1 and no R2.
- R2: This is the reverse read in two-ended sequencing. In two-ended sequencing, R1 and R2 are complementary read segments, derived from both ends of a DNA fragment.
- R3: It is rare and may occur in special experimental designs (such as three-terminal sequencing or more complex sequencing strategies). Usually used to represent additional read segments with multiple amplifications.
- I1: This is the first index read (also called barcode sequence). Indexed read segments are used to identify the source of a sample, often in multiplex sequencing.


## Use `cellranger-atac count` to preprocess fastq samples
The command line is as follows:
```
cellranger-atac count \
    --id=sampleID \
    --fastqs=00-RawData \
    --sample=sampleID \
    --reference=/l.ronghan/data/mm10/refdata-cellranger-arc-mm10-2020-A-2.0.0
```
There are a lot of files created in the output folder, including:
- web_summary.html - similar to gene expression
- summary.csv - similar to gene expression
- filtered_peak_bc_matrix - similar to gene expression
- filtered_tf_bc_matrix - similar to gene expression
- peak_annotation.tsv
- fragments.tsv.gz - peak barcode table
- singlecell.csv - barcodes summary
- BAMs - alignment files

## Use `cellranger-atac aggr`
```
cellranger-atac aggr \
  --id=combined \
  --csv=config.csv \
  --reference=/l.ronghan/data/mm10/refdata-cellranger-arc-mm10-2020-A-2.0.0 \
  --localcores=4 \
  --normalize=none \
  --dim-reduce=lsa \
  --localmem=4
```

